### PR TITLE
Deep clustering updates

### PR DESCRIPTION
**Pull request by sunits**
Deep clustering updates 
=> Create collate functions to ensure the seq length of batch elements are  the same
=> Bucketing sampler to ensure each element of the batch are approximately of the same seq length, this avoid chopping of large parts of the dataset
=> Fix the bug where different random samples were taken for mixture and sources
=> Python code to create wav id sample count files
=> Introduce VAD mask for deep clustering
=> Normalize the deep clustering loss
=> Fix bug in the chimera++ model where projection view mixes up the seq and source indexes
=> Take log of spectra as input to train the model
=> Simple eval script which needs to be enhanced
=> Training script without pytorch lightning (To be made compatible with the core asteroid)

### CODE REVIEW

File: asteroid/masknn/blocks.py [hunk 1]
```diff
@@ -446,9 +446,17 @@ def __init__(self, in_chan, n_src, rnn_type = 'lstm',
                 in_chan * embedding_dim)
         self.mask_layer = nn.Linear(rnn_out_dim, in_chan * n_src)
         self.non_linearity = nn.Sigmoid()
+        self.log = log
+        self.EPS = torch.finfo(torch.float32).eps
+        if log:
+            #TODO: Use pytorch lightning logger here
+            print('Using log spectrum as input')
```
Comment (mpariente) [inline asteroid/masknn/blocks.py:453]: It's fine not to print anything actually, it will be be printed in the conf dictionary. 
I would call it differently though, maybe `take_log` or `log_spec`?

File: asteroid/masknn/blocks.py [hunk 2]
```diff
@@ -433,7 +433,7 @@ class ChimeraPP(nn.Module):
     """
     def __init__(self, in_chan, n_src, rnn_type = 'lstm',
             embedding_dim=20, n_layers=2, hidden_size=600,
-            dropout=0, bidirectional=True):
+            dropout=0, bidirectional=True, log=False):
```
Comment (mpariente) [inline asteroid/masknn/blocks.py:436]: Update docstring as well

File: egs/wsj0-mix/DeepClustering/local/conf.yml [hunk 3]
```diff
@@ -1,8 +1,12 @@
...
-  n_filters: 512
-  kernel_size: 512
-  stride: 256
+# n_filters: 512
+# kernel_size: 512
+# stride: 256
+ n_filters: 256
+ kernel_size: 256
+ stride: 64
+ log: True # Use log spectra as input
```
Comment (mpariente) [inline egs/wsj0-mix/DeepClustering/local/conf.yml:9]: If the log is taken in the mask network, then it belongs in a mask network config

File: egs/wsj0-mix/DeepClustering/local/conf.yml [hunk 4]
```diff
@@ -18,13 +22,18 @@ optim:
...
-  wav_base_path: /srv/storage/talc3@talc-data.nancy/multispeech/calcul/users/ssivasankaran/experiments/data/speech_separation/wsj0-mix/2speakers_reverb_kinect_chime_noise_corrected/wav16k/max/
+    #  tr_wav_len_list: /srv/storage/talc3@talc-data.nancy/multispeech/calcul/users/ssivasankaran/experiments/data/speech_separation/wsj0-mix/2speakers_reverb_kinect_chime_noise_corrected/wav16k/max/tr.wavid.sample
+    #  cv_wav_len_list: /srv/storage/talc3@talc-data.nancy/multispeech/calcul/users/ssivasankaran/experiments/data/speech_separation/wsj0-mix/2speakers_reverb_kinect_chime_noise_corrected/wav16k/max/cv.wavid.sample
+    #  tt_wav_len_list: /srv/storage/talc3@talc-data.nancy/multispeech/calcul/users/ssivasankaran/experiments/data/speech_separation/wsj0-mix/2speakers_reverb_kinect_chime_noise_corrected/wav16k/max/tt.wavid.sample.head
+  tr_wav_len_list: exp/tr.wavid.samples
+  cv_wav_len_list: exp/cv.wavid.samples
+  tt_wav_len_list: exp/tt.wavid.samples
+  wav_base_path: /srv/storage/talc3@talc-data.nancy/multispeech/calcul/users/ssivasankaran/experiments/data/speech_separation/wsj0-mix/2speakers/wav8k/min/
```
Comment (mpariente) [inline egs/wsj0-mix/DeepClustering/local/conf.yml:31]: The config file shouldn't need modification, absolute path have to be in the run.sh. 
The text files containing infos about the dataset have to be under `./data/`

File: egs/wsj0-mix/DeepClustering/model.py [hunk 5]
```diff
@@ -33,8 +35,31 @@ def make_model_and_optimizer(conf):
     enc = fb.Encoder(fb.STFTFB(**conf['filterbank']))
     masker = ChimeraPP(int(enc.filterbank.n_feats_out/2), 2,
                        embedding_dim=20, n_layers=2, hidden_size=600, \
-                       dropout=0, bidirectional=True)
+                       dropout=0.5, bidirectional=True, \
+                       log=conf['filterbank']['log'])
```
Comment (mpariente) [inline egs/wsj0-mix/DeepClustering/model.py:39]: These constants should go in `conf.yml`, expose the ones that we'll experiment with in the run.sh as wel.

File: egs/wsj0-mix/DeepClustering/model.py [hunk 6]
```diff
@@ -33,8 +35,31 @@ def make_model_and_optimizer(conf):
...
+
+    Returns:
+        nn.Module the best pretrained model according to the val_loss.
+    """
+    # Create the model from recipe-local function
+    model, _ = make_model_and_optimizer(train_conf)
+    # Last best model summary
+    checkpoint = th.load(best_model_path, map_location='cpu')
+    # Removing additional saved info 
+    checkpoint['state_dict'].pop('enc.filterbank._filters')
```
Comment (mpariente) [inline egs/wsj0-mix/DeepClustering/model.py:60]: Why was this necessary? Did it crash?

File: egs/wsj0-mix/DeepClustering/model.py [hunk 7]
```diff
@@ -33,8 +35,31 @@ def make_model_and_optimizer(conf):
     enc = fb.Encoder(fb.STFTFB(**conf['filterbank']))
     masker = ChimeraPP(int(enc.filterbank.n_feats_out/2), 2,
                        embedding_dim=20, n_layers=2, hidden_size=600, \
-                       dropout=0, bidirectional=True)
+                       dropout=0.5, bidirectional=True, \
+                       log=conf['filterbank']['log'])
     model = Model(enc, masker)
```
Comment (mpariente) [inline egs/wsj0-mix/DeepClustering/model.py:40]: Even if the STFT is not inverted at train time, the goal with eventually be to go back to the time domain. You might as well attach the iSTFT to the model, it will make the rest easier IMO.

File: egs/wsj0-mix/DeepClustering/simple_train.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+            binary_mask=vad_tf_mask)
+    #loss = torch.sqrt(torch.pow(est_targets[1] - spec.permute(0,2,1), 2)).mean()
+    return dc_loss, loss
+
+def compute_vad(spectra, threshold_db=40):
+    ''' Compute a time-frequency VAD
+    source: github.com/funcwj/deep-clustering.git
+    '''
+    # to dB
+    spectra_db = 20 * torch.log10(spectra)
```
Comment (mpariente) [inline egs/wsj0-mix/DeepClustering/simple_train.py:144]: Stabilize log here

File: egs/wsj0-mix/DeepClustering/simple_train.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    max_magnitude_db = torch.max(spectra_db)
+    threshold = 10**((max_magnitude_db - threshold_db) / 20)
+    mask = spectra > threshold
+    return mask.double()
```
Comment (mpariente) [inline egs/wsj0-mix/DeepClustering/simple_train.py:148]: Is `float` not sufficient?

File: egs/wsj0-mix/create_meta_files.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import os
```
Comment (mpariente) [inline egs/wsj0-mix/create_meta_files.py:1]: Could this follow the same format as in the WHAM recipes? It would make it much easier to switch from one to the other (specially for the wsj0-3mix that is not covered in WHAM)

**Comment by mpariente** [comment]
Closed by #95 and #96
